### PR TITLE
Support underscores in class and definitions names in puppetParser

### PR DESCRIPTION
**Pull request by ahakanbaba**

### CODE REVIEW

File: Units/parser-puppetManifest.r/puppet-definitionname.d/expected.tags [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+one_	input.pp	/^define one_ {$/;"	definition	line:6	language:PuppetManifest	end:8
+/tmp/multipledefineone_	input.pp	/^    file { "\/tmp\/multipledefineone_": content => "one" }$/;"	resource	line:7	language:PuppetManifest	scope:definition:one_	end:7
+on_e	input.pp	/^define on_e {$/;"	definition	line:10	language:PuppetManifest	end:12
+/tmp/multipledefineon_e	input.pp	/^    file { "\/tmp\/multipledefineon_e": content => "one" }$/;"	resource	line:11	language:PuppetManifest	scope:definition:on_e	end:11
+o_ne::tw_o	input.pp	/^define o_ne::tw_o {$/;"	definition	line:15	language:PuppetManifest	end:17
+/tmp/multipledefineo_netw_o	input.pp	/^    file { "\/tmp\/multipledefineo_netw_o": content => "two" }$/;"	resource	line:16	language:PuppetManifest	scope:definition:o_ne::tw_o	end:16
+o_ne::two_	input.pp	/^define o_ne::two_ {$/;"	definition	line:19	language:PuppetManifest	end:21
+/tmp/multipledefineo_netwo_	input.pp	/^    file { "\/tmp\/multipledefineo_netwo_": content => "two" }$/;"	resource	line:20	language:PuppetManifest	scope:definition:o_ne::two_	end:20
+t	input.pp	/^t { 't': }$/;"	resource	line:23	language:PuppetManifest	end:23
```
Comment (ahakanbaba) [inline Units/parser-puppetManifest.r/puppet-definitionname.d/expected.tags]: This is probably another bug's indication. 
In the source file 5 different resources have been declared. 

```
t { 't': }
one_ { 'one_': }
on_e { 'on_e': }
o_ne::tw_o { 'o_ne::tw_o': }
o_ne::two_ { 'o_ne::two_': }
```
But this tag file only mentions one one them. 
I suspect there is another regular expression somewhere missing the "_" char. 

@masatake  if you point me to the right direction I can also fix that in another PR.

**Comment by ahakanbaba** [comment]
@masatake   I have added ontop of your commit [here](https://github.com/masatake/ctags/commit/9171923637c1d7f3effb72d44559f0ac1d60b07d)
Please take a look. 
In this PR I have renamed the test directories too. If you have a strong preference in the unit test directory names, I can rename them however you like.

**Comment by ahakanbaba** [comment]
Both of the new input.pp files are OK puppet syntax: 

```
puppet-classname.d (master)]$ puppet apply --noop input.pp
Notice: Compiled catalog for hbaba.dev.box.net in environment production in 0.12 seconds
Notice: /Stage[main]/T/File[/tmp/multipleclasst]/ensure: current_value absent, should be file (noop)
Notice: Class[T]: Would have triggered 'refresh' from 1 events
Notice: /Stage[main]/One_/File[/tmp/multipleclassone_]/ensure: current_value absent, should be file (noop)
Notice: Class[One_]: Would have triggered 'refresh' from 1 events
Notice: /Stage[main]/On_e/File[/tmp/multipleclasson_e]/ensure: current_value absent, should be file (noop)
Notice: Class[On_e]: Would have triggered 'refresh' from 1 events
Notice: /Stage[main]/O_ne::Tw_o/File[/tmp/multipleclasso_netw_o]/ensure: current_value absent, should be file (noop)
Notice: Class[O_ne::Tw_o]: Would have triggered 'refresh' from 1 events
Notice: /Stage[main]/O_ne::Two_/File[/tmp/multipleclasso_netwo_]/ensure: current_value absent, should be file (noop)
Notice: Class[O_ne::Two_]: Would have triggered 'refresh' from 1 events
Notice: Stage[main]: Would have triggered 'refresh' from 5 events
Notice: Applied catalog in 0.11 seconds
```

and: 

```
puppet-definitionname.d (master)]$ puppet apply --noop input.pp
Notice: Compiled catalog for hbaba.dev.box.net in environment production in 0.10 seconds
Notice: /Stage[main]/Main/T[t]/File[/tmp/multipledefinet]/ensure: current_value absent, should be file (noop)
Notice: T[t]: Would have triggered 'refresh' from 1 events
Notice: /Stage[main]/Main/One_[one_]/File[/tmp/multipledefineone_]/ensure: current_value absent, should be file (noop)
Notice: One_[one_]: Would have triggered 'refresh' from 1 events
Notice: /Stage[main]/Main/On_e[on_e]/File[/tmp/multipledefineon_e]/ensure: current_value absent, should be file (noop)
Notice: On_e[on_e]: Would have triggered 'refresh' from 1 events
Notice: /Stage[main]/Main/O_ne::Tw_o[o_ne::tw_o]/File[/tmp/multipledefineo_netw_o]/ensure: current_value absent, should be file (noop)
Notice: O_ne::Tw_o[o_ne::tw_o]: Would have triggered 'refresh' from 1 events
Notice: /Stage[main]/Main/O_ne::Two_[o_ne::two_]/File[/tmp/multipledefineo_netwo_]/ensure: current_value absent, should be file (noop)
Notice: O_ne::Two_[o_ne::two_]: Would have triggered 'refresh' from 1 events
Notice: Class[Main]: Would have triggered 'refresh' from 5 events
Notice: Stage[main]: Would have triggered 'refresh' from 1 events
Notice: Applied catalog in 0.08 seconds
```

**Comment by ahakanbaba** [comment]
@masatake  Could you take a look at this at your convenience please ?

**Comment by masatake** [comment]
I got following output for the input
```
[jet@localhost]~/var/ctags% ./ctags  --sort=no -o - foo.pp
t	foo.pp	/^t { 't': }$/;"	r
one_	foo.pp	/^one_ { 'one_': }$/;"	r
on_e	foo.pp	/^on_e { 'on_e': }$/;"	r
o_ne::tw_o	foo.pp	/^o_ne::tw_o { 'o_ne::tw_o': }$/;"	r
o_ne::two_	foo.pp	/^o_ne::two_ { 'o_ne::two_': }$/;"	r
```
wiith following change:
```
diff --git a/optlib/puppetManifest.ctags b/optlib/puppetManifest.ctags
index c5ab0c7c..ccb0254b 100644
--- a/optlib/puppetManifest.ctags
+++ b/optlib/puppetManifest.ctags
@@ -152,8 +152,8 @@
 #
 # block
 #
---_mtable-regex-PuppetManifest=blockStart/@?::[a-zA-Z0-9:]+[ \t\n]*\{//{tenter=resourceBlock}
---_mtable-regex-PuppetManifest=blockStart/@?[a-zA-Z][a-zA-Z0-9:]*[ \t\n]*\{//{tenter=resourceBlock}
+--_mtable-regex-PuppetManifest=blockStart/@?::[a-zA-Z0-9:_]+[ \t\n]*\{//{tenter=resourceBlock}
+--_mtable-regex-PuppetManifest=blockStart/@?[a-zA-Z][a-zA-Z0-9:_]*[ \t\n]*\{//{tenter=resourceBlock}
```

**Comment by masatake** [comment]
By the way, it is a really bad idea using `master` branch as the "from" branch for making a pull request.
See https://git-scm.com/book/en/v2/Git-Branching-Branching-Workflows

**Comment by masatake** [comment]
Could you put "PuppetManifest:" as the prefix for the commit logs?

**Comment by ahakanbaba** [comment]
> By the way, it is a really bad idea using `master` branch as the "from" branch for making a pull request.
> See https://git-scm.com/book/en/v2/Git-Branching-Branching-Workflows

You are right. I will not use ahakanbaba:master again.

**Comment by ahakanbaba** [comment]
Closed in favor of #1906  
There the commits are squashed and proper branch naming is used. 
I will carry the open conversations from here to #1906

**Pull request closed without merging**